### PR TITLE
App 1431 core allow multisig to restrict proposal creation to signers or anyone

### PR DESCRIPTION
**Pull request by mathewmeconry**
## Description

Adds a new setting to the Multisig plugin to allow `onlyListed` users to create proposals.

Depends on #167 

Task: https://aragonassociation.atlassian.net/browse/APP-1431

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I ran all tests with success and extended them if necessary.
- [x] I have updated the ``CHANGELOG.md`` file in the root folder.
- [ ] I have tested my code on the test network.

[APP-1431]: https://aragonassociation.atlassian.net/browse/APP-1431?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/contracts/contracts/voting/multisig/Multisig.sol [hunk 1]
```diff
@@ -49,6 +49,12 @@ contract Multisig is TimeHelpers, PluginUUPSUpgradeable, Addresslist {
         uint256 addresslistLength;
     }
 
+    /// @notice A container for the plugin settings
+    /// @param onlyListed Whether only members in the addresslist can create a proposal
+    struct PluginSettings {
+        bool onlyListed;
```
Comment (novaknole) [inline packages/contracts/contracts/voting/multisig/Multisig.sol:55]: It seems like `minApprovals_` is also `settings` kind of guy.  Maybe we can put them in the same struct ? Let's reconfirm with @Michael-A-Heuer

File: packages/contracts/contracts/voting/multisig/Multisig.sol [hunk 2]
```diff
@@ -213,7 +238,7 @@ contract Multisig is TimeHelpers, PluginUUPSUpgradeable, Addresslist {
     ) external returns (uint256 proposalId) {
         uint64 snapshotBlock = getBlockNumber64() - 1;
 
-        if (!isListedAtBlock(_msgSender(), snapshotBlock)) {
+        if (!isListedAtBlock(_msgSender(), snapshotBlock) && pluginSettings.onlyListed) {
```
[the remark this answers is not in the document]
Comment (mathewmeconry) [reply packages/contracts/contracts/voting/multisig/Multisig.sol:241]: done

File: packages/contracts/contracts/voting/multisig/Multisig.sol [hunk 3]
```diff
@@ -73,6 +79,9 @@ contract Multisig is TimeHelpers, PluginUUPSUpgradeable, Addresslist {
     /// @notice A mapping between proposal IDs and proposal information.
     mapping(uint256 => Proposal) internal proposals;
 
+    /// @notice The current plugin settings
+    PluginSettings pluginSettings;
```
[the remark this answers is not in the document]
Comment (mathewmeconry) [reply packages/contracts/contracts/voting/multisig/Multisig.sol:83]: done

File: packages/contracts/contracts/voting/multisig/Multisig.sol (hunk 1 above)
Comment (mathewmeconry) [reply packages/contracts/contracts/voting/multisig/Multisig.sol:55]: good point. @Michael-A-Heuer  wdyt?

File: packages/contracts/contracts/voting/multisig/Multisig.sol [hunk 4]
```diff
@@ -49,6 +49,12 @@ contract Multisig is TimeHelpers, PluginUUPSUpgradeable, Addresslist {
         uint256 addresslistLength;
     }
 
+    /// @notice A container for the plugin settings
+    /// @param onlyListed Whether only members in the addresslist can create a proposal
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/voting/multisig/Multisig.sol:53]:
```suggestion
    /// @notice A container for the plugin settings.
    /// @param onlyListed Whether only members in the addresslist can create a proposal.
```

File: packages/contracts/contracts/voting/multisig/Multisig.sol (hunk 1 above)
Comment (Michael-A-Heuer) [reply packages/contracts/contracts/voting/multisig/Multisig.sol:55]: Yes, it should also be in the struct. In `MajorityVotingBase`, we named it 
`struct VotingSettings`. 
Should it be named `struct ApprovalSettings` here?

File: packages/contracts/contracts/voting/multisig/Multisig.sol [hunk 5]
```diff
@@ -73,6 +79,9 @@ contract Multisig is TimeHelpers, PluginUUPSUpgradeable, Addresslist {
     /// @notice A mapping between proposal IDs and proposal information.
     mapping(uint256 => Proposal) internal proposals;
 
+    /// @notice The current plugin settings
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/voting/multisig/Multisig.sol:82]:
```suggestion
    /// @notice The current plugin settings.
```

File: packages/contracts/contracts/voting/multisig/Multisig.sol [hunk 6]
```diff
@@ -120,21 +129,28 @@ contract Multisig is TimeHelpers, PluginUUPSUpgradeable, Addresslist {
     /// @param execResults The bytes array resulting from the proposal execution in the associated DAO.
     event ProposalExecuted(uint256 indexed proposalId, bytes[] execResults);
 
+    /// @notice Emitted when the plugin settings are set
+    /// @param onlyListed Whether only members in the addresslist can create a proposal
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/voting/multisig/Multisig.sol:133]:
```suggestion
    /// @notice Emitted when the plugin settings are set.
    /// @param onlyListed Whether only listed addresses can create a proposal.
```

**Review by brickpop — APPROVED** [review]

**Review by Michael-A-Heuer — COMMENTED** [review]
Minor suggestions and comments related to naming. Apart from that, the PR is good to go

**Review by Michael-A-Heuer — CHANGES_REQUESTED** [review]
Please make sure to adapt the storage gap. I think `uint256[46] private __gap;` would be correct here.

File: packages/contracts/contracts/voting/multisig/Multisig.sol [hunk 7]
```diff
@@ -120,21 +129,28 @@ contract Multisig is TimeHelpers, PluginUUPSUpgradeable, Addresslist {
...
+    event PluginSettingsUpdated(bool onlyListed);
+
     /// @notice Initializes the component.
     /// @dev This method is required to support [ERC-1822](https://eips.ethereum.org/EIPS/eip-1822).
     /// @param _dao The IDAO interface of the associated DAO.
     // /// @param _majorityVotingSettings The majority voting settings.
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/voting/multisig/Multisig.sol:139]:
```suggestion
    /// @param _plugingSettings The multisig settings.
```

File: packages/contracts/contracts/voting/multisig/Multisig.sol [hunk 8]
```diff
@@ -49,6 +49,12 @@ contract Multisig is TimeHelpers, PluginUUPSUpgradeable, Addresslist {
         uint256 addresslistLength;
     }
 
+    /// @notice A container for the plugin settings
+    /// @param onlyListed Whether only members in the addresslist can create a proposal
+    struct PluginSettings {
```
[the remark this answers is not in the document]
Comment (mathewmeconry) [reply packages/contracts/contracts/voting/multisig/Multisig.sol]: renamed

File: packages/contracts/test/voting/multisig.ts [hunk 9]
```diff
@@ -1,11 +1,11 @@
...
+import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers';
 
-import {DAO} from '../../typechain';
-import {findEvent, DAO_EVENTS, VOTING_EVENTS} from '../../utils/event';
-import {getMergedABI} from '../../utils/abi';
-import {customError, ERRORS} from '../test-utils/custom-error-helper';
+import { DAO } from '../../typechain';
+import { findEvent, DAO_EVENTS, VOTING_EVENTS } from '../../utils/event';
+import { getMergedABI } from '../../utils/abi';
+import { customError, ERRORS } from '../test-utils/custom-error-helper';
```
[the remark this answers is not in the document]
Comment (mathewmeconry) [reply packages/contracts/test/voting/multisig.ts]: I use it now

**Comment by brickpop** [comment]
Looks good to me, but I'd wait for @novaknole and @Michael-A-Heuer to share their thoughts :+1:

**Review by Michael-A-Heuer — APPROVED** [review]

**Review by Michael-A-Heuer — APPROVED** [review]

**Pull request merged**